File: flight_api/app/data/mock_data.py
"""
In-memory mock data store. Treated as the authoritative database for the API.
"""

from copy import deepcopy
from typing import Any

_FLIGHTS_RAW: list[dict[str, Any]] = [
    {
        "id": "FL001",
        "flight_number": "EK203",
        "airline": "Emirates",
        "aircraft_type": "A380",
        "aircraft_size": "F",
        "origin": "LHR",
        "destination": "DXB",
        "terminal": "T1",
        "assigned_stand": "A1-01",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T06:30:00Z",
        "estimated_time": "2025-01-15T06:25:00Z",
        "block_time_start": "2025-01-15T06:30:00Z",
        "block_time_end": "2025-01-15T08:45:00Z",
        "status": "on_time",
        "pax_count": 489,
    },
    {
        "id": "FL002",
        "flight_number": "QR501",
        "airline": "Qatar Airways",
        "aircraft_type": "B777-300ER",
        "aircraft_size": "E",
        "origin": "DOH",
        "destination": "DXB",
        "terminal": "T1",
        "assigned_stand": "A1-01",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T10:15:00Z",
        "estimated_time": "2025-01-15T10:40:00Z",
        "block_time_start": "2025-01-15T10:15:00Z",
        "block_time_end": "2025-01-15T12:30:00Z",
        "status": "delayed",
        "pax_count": 356,
    },
    {
        "id": "FL003",
        "flight_number": "BA107",
        "airline": "British Airways",
        "aircraft_type": "B787-9",
        "aircraft_size": "E",
        "origin": "LHR",
        "destination": "DXB",
        "terminal": "T1",
        "assigned_stand": "A1-02",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T07:00:00Z",
        "estimated_time": "2025-01-15T07:00:00Z",
        "block_time_start": "2025-01-15T07:00:00Z",
        "block_time_end": "2025-01-15T09:15:00Z",
        "status": "on_time",
        "pax_count": 264,
    },
    {
        "id": "FL004",
        "flight_number": "LH752",
        "airline": "Lufthansa",
        "aircraft_type": "A340-600",
        "aircraft_size": "E",
        "origin": "FRA",
        "destination": "DXB",
        "terminal": "T1",
        "assigned_stand": "A1-02",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T11:30:00Z",
        "estimated_time": "2025-01-15T11:30:00Z",
        "block_time_start": "2025-01-15T11:30:00Z",
        "block_time_end": "2025-01-15T14:00:00Z",
        "status": "on_time",
        "pax_count": 291,
    },
    {
        "id": "FL005",
        "flight_number": "SQ321",
        "airline": "Singapore Airlines",
        "aircraft_type": "A350-900",
        "aircraft_size": "E",
        "origin": "SIN",
        "destination": "DXB",
        "terminal": "T1",
        "assigned_stand": "A1-03",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T08:45:00Z",
        "estimated_time": "2025-01-15T08:50:00Z",
        "block_time_start": "2025-01-15T08:45:00Z",
        "block_time_end": "2025-01-15T11:00:00Z",
        "status": "on_time",
        "pax_count": 303,
    },
    {
        "id": "FL006",
        "flight_number": "EK512",
        "airline": "Emirates",
        "aircraft_type": "B777-300ER",
        "aircraft_size": "E",
        "origin": "DXB",
        "destination": "JFK",
        "terminal": "T1",
        "assigned_stand": "A1-03",
        "operation": "departure",
        "scheduled_time": "2025-01-15T13:00:00Z",
        "estimated_time": "2025-01-15T13:00:00Z",
        "block_time_start": "2025-01-15T11:30:00Z",
        "block_time_end": "2025-01-15T13:00:00Z",
        "status": "on_time",
        "pax_count": 396,
    },
    {
        "id": "FL007",
        "flight_number": "QF001",
        "airline": "Qantas",
        "aircraft_type": "A380",
        "aircraft_size": "F",
        "origin": "SYD",
        "destination": "DXB",
        "terminal": "T2",
        "assigned_stand": "B1-01",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T05:30:00Z",
        "estimated_time": "2025-01-15T05:20:00Z",
        "block_time_start": "2025-01-15T05:30:00Z",
        "block_time_end": "2025-01-15T08:00:00Z",
        "status": "early",
        "pax_count": 450,
    },
    {
        "id": "FL008",
        "flight_number": "CX888",
        "airline": "Cathay Pacific",
        "aircraft_type": "A350-1000",
        "aircraft_size": "E",
        "origin": "HKG",
        "destination": "DXB",
        "terminal": "T2",
        "assigned_stand": "B1-01",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T10:00:00Z",
        "estimated_time": "2025-01-15T09:50:00Z",
        "block_time_start": "2025-01-15T10:00:00Z",
        "block_time_end": "2025-01-15T12:30:00Z",
        "status": "early",
        "pax_count": 334,
    },
    {
        "id": "FL009",
        "flight_number": "JL043",
        "airline": "Japan Airlines",
        "aircraft_type": "B787-9",
        "aircraft_size": "E",
        "origin": "NRT",
        "destination": "DXB",
        "terminal": "T2",
        "assigned_stand": "B1-02",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T14:30:00Z",
        "estimated_time": "2025-01-15T14:30:00Z",
        "block_time_start": "2025-01-15T14:30:00Z",
        "block_time_end": "2025-01-15T16:45:00Z",
        "status": "on_time",
        "pax_count": 186,
    },
    {
        "id": "FL010",
        "flight_number": "AF218",
        "airline": "Air France",
        "aircraft_type": "A330-200",
        "aircraft_size": "D",
        "origin": "CDG",
        "destination": "DXB",
        "terminal": "T2",
        "assigned_stand": "B1-03",
        "operation": "arrival",
        "scheduled_time": "2025-01-15T16:00:00Z",
        "estimated_time": "2025-01-15T16:15:00Z",
        "block_time_start": "2025-01-15T16:00:00Z",
        "block_time_end": "2025-01-15T18:30:00Z",
        "status": "delayed",
        "pax_count": 224,
    },
]

_STANDS_RAW: list[dict[str, Any]] = [
    {"id": "A1-01", "terminal": "T1", "type": "contact", "has_plb": True, "max_aircraft_size": "F", "associated_gate": "G01", "position": {"x": 100, "y": 50}},
    {"id": "A1-02", "terminal": "T1", "type": "contact", "has_plb": True, "max_aircraft_size": "E", "associated_gate": "G02", "position": {"x": 100, "y": 120}},
    {"id": "A1-03", "terminal": "T1", "type": "contact", "has_plb": True, "max_aircraft_size": "E", "associated_gate": "G03", "position": {"x": 100, "y": 190}},
    {"id": "A1-04", "terminal": "T1", "type": "contact", "has_plb": True, "max_aircraft_size": "D", "associated_gate": "G04", "position": {"x": 100, "y": 260}},
    {"id": "A1-05", "terminal": "T1", "type": "remote", "has_plb": False, "max_aircraft_size": "E", "associated_gate": None, "position": {"x": 100, "y": 330}},
    {"id": "B1-01", "terminal": "T2", "type": "contact", "has_plb": True, "max_aircraft_size": "F", "associated_gate": "G05", "position": {"x": 400, "y": 50}},
    {"id": "B1-02", "terminal": "T2", "type": "contact", "has_plb": True, "max_aircraft_size": "E", "associated_gate": "G06", "position": {"x": 400, "y": 120}},
    {"id": "B1-03", "terminal": "T2", "type": "contact", "has_plb": False, "max_aircraft_size": "D", "associated_gate": "G07", "position": {"x": 400, "y": 190}},
    {"id": "B1-04", "terminal": "T2", "type": "remote", "has_plb": False, "max_aircraft_size": "F", "associated_gate": None, "position": {"x": 400, "y": 260}},
    {"id": "B1-05", "terminal": "T2", "type": "remote", "has_plb": False, "max_aircraft_size": "D", "associated_gate": None, "position": {"x": 400, "y": 330}},
]


class DataStore:
    """Singleton-like in-memory store. Mutations persist for the lifetime of the process."""

    def __init__(self) -> None:
        self._flights: list[dict[str, Any]] = deepcopy(_FLIGHTS_RAW)
        self._stands: list[dict[str, Any]] = deepcopy(_STANDS_RAW)

    # ------------------------------------------------------------------
    # Flights
    # ------------------------------------------------------------------

    def get_flights(self) -> list[dict[str, Any]]:
        return self._flights

    def get_flight_by_id(self, flight_id: str) -> dict[str, Any] | None:
        return next((f for f in self._flights if f["id"] == flight_id), None)

    def update_flight(self, flight_id: str, updates: dict[str, Any]) -> dict[str, Any] | None:
        for i, f in enumerate(self._flights):
            if f["id"] == flight_id:
                self._flights[i] = {**f, **updates}
                return self._flights[i]
        return None

    # ------------------------------------------------------------------
    # Stands
    # ------------------------------------------------------------------

    def get_stands(self) -> list[dict[str, Any]]:
        return self._stands

    def get_stand_by_id(self, stand_id: str) -> dict[str, Any] | None:
        return next((s for s in self._stands if s["id"] == stand_id), None)


# Module-level singleton — imported by the service layer
store = DataStore()

File: client/src/Resource/data/mockData.ts
import type { MockData } from "../types";

/**
 * MOCK_DATA — static seed data representing the airport resource schema.
 * Swap this out for an API fetch (e.g. `useSWR('/api/resource-graph')`)
 * without touching any component or hook.
 */
export const MOCK_DATA: MockData = {
  flights: [
    { id: "FL001", flight_number: "EK203",  aircraft_type: "A380",        assigned_stand: "A1-01", status: "on_time", estimated_time: "2025-01-15T06:25:00Z" },
    { id: "FL002", flight_number: "QR501",  aircraft_type: "B777-300ER",  assigned_stand: "A1-01", status: "delayed", estimated_time: "2025-01-15T10:40:00Z" },
    { id: "FL003", flight_number: "BA107",  aircraft_type: "B787-9",      assigned_stand: "A1-02", status: "on_time", estimated_time: "2025-01-15T07:00:00Z" },
    { id: "FL004", flight_number: "LH752",  aircraft_type: "A340-600",    assigned_stand: "A1-02", status: "on_time", estimated_time: "2025-01-15T11:30:00Z" },
    { id: "FL005", flight_number: "SQ321",  aircraft_type: "A350-900",    assigned_stand: "A1-03", status: "on_time", estimated_time: "2025-01-15T08:50:00Z" },
    { id: "FL006", flight_number: "EK512",  aircraft_type: "B777-300ER",  assigned_stand: "A1-03", status: "on_time", estimated_time: "2025-01-15T13:00:00Z" },
    { id: "FL007", flight_number: "QF001",  aircraft_type: "A380",        assigned_stand: "B1-01", status: "early",   estimated_time: "2025-01-15T05:20:00Z" },
    { id: "FL008", flight_number: "CX888",  aircraft_type: "A350-1000",   assigned_stand: "B1-01", status: "early",   estimated_time: "2025-01-15T09:50:00Z" },
    { id: "FL009", flight_number: "JL043",  aircraft_type: "B787-9",      assigned_stand: "B1-02", status: "on_time", estimated_time: "2025-01-15T14:30:00Z" },
    { id: "FL010", flight_number: "AF218",  aircraft_type: "A330-200",    assigned_stand: "B1-03", status: "delayed", estimated_time: "2025-01-15T16:15:00Z" },
  ],
  stands: [
    { id: "A1-01", terminal: "T1", type: "contact", max_aircraft_size: "F", position: { x: 100, y: 50  } },
    { id: "A1-02", terminal: "T1", type: "contact", max_aircraft_size: "E", position: { x: 100, y: 150 } },
    { id: "A1-03", terminal: "T1", type: "contact", max_aircraft_size: "E", position: { x: 100, y: 250 } },
    { id: "A1-04", terminal: "T1", type: "contact", max_aircraft_size: "D", position: { x: 100, y: 350 } },
    { id: "A1-05", terminal: "T1", type: "remote",  max_aircraft_size: "E", position: { x: 100, y: 450 } },
    { id: "B1-01", terminal: "T2", type: "contact", max_aircraft_size: "F", position: { x: 600, y: 50  } },
    { id: "B1-02", terminal: "T2", type: "contact", max_aircraft_size: "E", position: { x: 600, y: 150 } },
    { id: "B1-03", terminal: "T2", type: "contact", max_aircraft_size: "D", position: { x: 600, y: 250 } },
    { id: "B1-04", terminal: "T2", type: "remote",  max_aircraft_size: "F", position: { x: 600, y: 350 } },
    { id: "B1-05", terminal: "T2", type: "remote",  max_aircraft_size: "D", position: { x: 600, y: 450 } },
  ],
  gates: [
    { id: "G01", terminal: "T1", type: "contact",     position: { x: 300, y: 50  } },
    { id: "G02", terminal: "T1", type: "contact",     position: { x: 300, y: 150 } },
    { id: "G03", terminal: "T1", type: "contact",     position: { x: 300, y: 300 } },
    { id: "G04", terminal: "T1", type: "contact",     position: { x: 300, y: 450 } },
    { id: "G05", terminal: "T2", type: "contact",     position: { x: 800, y: 50  } },
    { id: "G06", terminal: "T2", type: "contact",     position: { x: 800, y: 150 } },
    { id: "G07", terminal: "T2", type: "non_contact", position: { x: 800, y: 350 } },
  ],
  graph_edges: {
    plb_connections: [
      { stand: "A1-01", gate: "G01", type: "plb" },
      { stand: "A1-02", gate: "G02", type: "plb" },
      { stand: "A1-03", gate: "G03", type: "plb" },
      { stand: "A1-04", gate: "G04", type: "plb" },
      { stand: "B1-01", gate: "G05", type: "plb" },
      { stand: "B1-02", gate: "G06", type: "plb" },
    ],
    walking_connections: [
      { stand: "A1-04", gate: "G03", type: "walking" },
      { stand: "B1-03", gate: "G07", type: "walking" },
      { stand: "B1-04", gate: "G07", type: "bus"     },
      { stand: "B1-05", gate: "G07", type: "bus"     },
    ],
    adjacency_constraints: [
      { stand_a: "A1-01", stand_b: "A1-02", type: "wingtip"   },
      { stand_a: "A1-03", stand_b: "A1-04", type: "wingtip"   },
      { stand_a: "B1-01", stand_b: "B1-02", type: "adjacency" },
    ],
  },
};